tensorboard/loader.py

```diff
@@ -399,6 +401,254 @@ def __str__(self):
+          since the UNIX epoch in zulu time.
+      sleep: Injected time.sleep function.
+      log_callback: Callback for emitting normal log records.
+      bar_callback: Callback for emitting ephemeral bar records.
+      rate_counter_factory: Constructor to RateCounter, which can be
+          swapped out for testing.
+
+    :type clock: () -> float
+    :type sleep: (float) -> None
+    :type rate_counter_factory: (float) -> RateCounter
+    """
+    self._clock = clock
+    self._sleep = sleep
+    self._log_callback = log_callback
+    self._bar_callback = bar_callback
+    self._initialized = False
+    self._offset = 0
+    self._size = 0
+    self._last_log_time = 0.0
+    self._last_bar_time = 0.0
+    self._rate_offset = rate_counter_factory(Progress.RATE_WINDOW)
+    self._rate_size = rate_counter_factory(Progress.RATE_WINDOW)
+
+  def set_progress(self, offset, size):
+    """Updates the progress bar state.
+
+    This method will cause progress information to be occasionally
+    written out.
+
+    Args:
+      offset: The number of bytes processed so far.
+      size: The total number of bytes. This is allowed to increase or
+          decrease, but it must remain greater than offset.
+
+    Raises:
+      ValueError: If offset is greater than size, or offset or size
+          decreased from the last invocation.
+
+    :type offset: int
+    :type size: int
+    """
+    if offset > size:
+      raise ValueError('offset (%d) can not exceed size (%d)' % (offset, size))
+    self._rate_offset.set_value(offset)
+    self._rate_size.set_value(size)
+    self._offset = offset
+    self._size = size
+    now = self._clock()
+    if not self._initialized:
+      self._last_log_time = now
+      self._last_bar_time = now
+      self._initialized = True
+      return
+    elapsed = now - self._last_log_time
+    if elapsed >= Progress.LOG_INTERVAL_SECONDS:
+      self._last_log_time = now
+      self._show_log()
+    elapsed = now - self._last_bar_time
+    if elapsed >= Progress.BAR_INTERVAL_SECONDS:
+      self._last_bar_time = now
+      self._show_bar()
+
+  def close(self):
+    """Forces progress to be written to log.
+
+    This method exists because we don't want the progress bar to say
+    something like 98% once the file is done loading.
+    """
+    self._show_log()
+    self._show_bar()
+    # Instructs util.LogHandler to clear the ephemeral logging state.
+    self._bar_callback('')
+
+  def sleep(self, seconds):
+    """Sleeps for a given number of seconds.
+
+    Time spent sleeping in this method does not have a detrimental
+    impact on the consumption rate.
+
+    :type seconds: float
+    """
+    self._sleep(seconds)
+    self._rate_offset.bump()
+
+  def _show_log(self):
+    self._log_callback('Loaded %s', self._get_message())
+
+  def _show_bar(self):
+    sofar = int(self._get_fraction() * Progress.BAR_WIDTH)
+    bar = (Progress.BLOCK_DARK * sofar +
```

Comment:
Amusingly, the linter marks this as a "blacklisted name." Feel free to `# pylint: disable=blacklisted-name` for this function. :-)